src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -2238,73 +1981,107 @@ tx_gateway_frontend::do_commit_tm_tx(
         co_return tx::errc::not_coordinator;
     }
 
-    if (tx.status == tx_status::completed_commit) {
-        outcome->set_value(tx::errc::none);
-        co_return co_await recommit_tm_tx(stm, expected_term, tx, timeout);
+    if (tx.status == tx_status::ongoing || tx.status == tx_status::empty) {
+        auto update_result = co_await stm->update_transaction_status(
+          expected_term, tx.id, tx_status::preparing_commit);
+        if (!update_result) {
+            auto err = map_state_update_outcome(update_result.error());
+            outcome->set_value(err);
+            co_return err;
+        }
     }
 
+    auto commit_result = co_await commit_data(stm, expected_term, tx, timeout);
+    if (commit_result.has_error()) {
+        outcome->set_value(commit_result.error());
+        co_return commit_result.error();
+    }
     vlog(
       txlog.trace,
-      "[tx_id={}] marking transaction pid: {} tx_seq: {} etag: {} as prepared "
-      "in term: {}",
+      "[tx_id={}] marking transaction {} as commit completed in term: {}",
       tx.id,
-      tx.pid,
-      tx.tx_seq,
-      tx.etag,
+      tx,
       expected_term);
-    auto changed_tx = co_await stm->mark_tx_prepared(expected_term, tx.id);
-    if (!changed_tx.has_value()) {
+    auto committed_tx = co_await stm->finish_transaction(
+      expected_term, tx.id, tx_status::completed_commit);
+    if (!committed_tx.has_value()) {
         vlog(
           txlog.trace,
-          "[tx_id={}] error committing transaction pid: {} etag: {} tx_seq: {} "
-          "- {}",
+          "[tx_id={}] error committing transaction: {} - {}",
           tx.id,
-          tx.pid,
-          tx.etag,
-          tx.tx_seq,
-          changed_tx.error());
+          tx,
+          committed_tx.error());
+        auto err = map_state_update_outcome(committed_tx.error());
 
-        if (!is_fetch_tx_supported()) {
-            if (changed_tx.error() == tm_stm::op_status::not_leader) {
-                outcome->set_value(tx::errc::not_coordinator);
-                co_return tx::errc::not_coordinator;
-            }
+        outcome->set_value(err);
+        co_return err;
+    }
+    outcome->set_value(tx::errc::none);
```

Comment:
I think there is a behavioral difference here in the 2PC, we are waiting for commit_data() to return and then ack-ing the client, doesn't it have to be before L1994